templates/terraform/objectlib/base.go.erb

```diff
@@ -6,15 +6,21 @@ package google
     resource_name = product_ns + object.name
     properties = object.all_user_properties
     api_version = version.base_url.split("/")[-1]
+    base_url_name = object.base_url + '/{{name}}'
+    asset_name_template = '//' + product_ns.downcase + '.googleapis.com/' + (object.has_self_link ? base_url_name : object.self_link || base_url_name)
```

Comment:
This logic doesn't seem right to me. If the object has a self_link then you use the base_url? I would have expected `(object.has_self_link ? object.self_link : base_url_name)`

Reply:
@rileykarson

Reply:
Ah, didn't see this- confusingly, `has_self_link` means the object has a `self_link` field, which is always `base_url/{{name}}`. Also confusingly, `base_url` isn't really a url- it's only part of one, appended on an actual `base_url`. Resources with `name` fields have explicit self links, set by `self_link`

Reply:
Regardless of that, I still don't think this logic is correct. If `has_self_link` is false, you will end up with the value of  `object.self_link || base_url_name` being assigned to your `asset_name_template`. If `object.self_link` is an empty string when `has_self_link` is false (and I really hope it is) and given that ruby evaluates the empty string as truthy then `object.self_link || "a_real_url"` will evaluate to the empty string.

I checked the products being generated by `compute` and right now all of them have `has_self_link` set to true, so the downstreams wouldn't show this problem either. 

If there is no actual relationship between `self_link` and `has_self_link` as this code seems to imply, then we should probably check the contents of `self_link` to determine whether to use it. 

And if this is actually working as intended in the `false` scenario, then at minimum there should be a comment explaining what's happening here. At first glance it looks like the ternary got accidentally inverted and somebody will probably "fix" it one day while cleaning up code.

Reply:
Here are a few scenarios that this is coded for:

File: `products/compute/api.yaml`
Goal: `//compute.googleapis.com/projects/{project}/zones/{zone}/disks/{disk}`
Logic: `has_self_link == true ... want base_url (with name appended)`

```yaml
    name: 'Disk'
    base_url: projects/{{project}}/zones/{{zone}}/disks
    has_self_link: true
    # No self_link
```

File: `products/redis/api.yaml`
Goal: `//redis.googleapis.com/projects/{project_id}/locations/{location_id}/instances/{instance_id}`
Logic: `has_self_link == <empty> && self_link is defined ... want self_link`

```yaml
    name: 'Instance'
    base_url: |
      projects/{{project}}/locations/{{region}}/instances?instanceId={{name}}
    self_link: projects/{{project}}/locations/{{region}}/instances/{{name}}
    # No has_self_link
```

File: `products/pubsub/api.yaml`
Goal: `//pubsub.googleapis.com/projects/{project}/topics/{topic}`
Logic: `has_self_link == <empty> && self_link == <empty>... want base_url (with name appended)`

```yaml
    name: 'Topic'
    base_url: projects/{{project}}/topics
    # No self_link, has_self_link
```

For the goals, see: https://cloud.google.com/resource-manager/docs/cloud-asset-inventory/resource-name-format

Reply:
Well I'm sad to discover that it does work as intended. But given that `has_self_link` has no relation with `self_link` as I found by generating some of the permutations:
```
build/google/spanner_instance.go:	// self_link_investigate has_self_link[false] self_link[] base_url[projects/{{project}}/instances]
build/google/compute_autoscaler.go:	// self_link_investigate has_self_link[true] self_link[] base_url[projects/{{project}}/zones/{{zone}}/autoscalers]
```
I think you should ignore that field entirely as it's misleading at best and will hopefully be refactored in the near future. This code shouldn't have to understand the legacy workings of MM and in the future I imagine that `has_self_link` will be removed and `self_link` will either be populated or not. You'll be future proofed by just checking for the existence of `self_link`  which should be correct (if it isn't then it would be a bug to fix).

Reply:
Good now?

Reply:
Much less scary, thanks!